components/service/sync-logins/src/test/java/mozilla/components/service/sync/logins/DefaultLoginValidationDelegateTest.kt

```diff
@@ -0,0 +1,203 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package mozilla.components.service.sync.logins
+
+import androidx.test.ext.junit.runners.AndroidJUnit4
+import kotlinx.coroutines.ExperimentalCoroutinesApi
+import kotlinx.coroutines.test.TestCoroutineScope
+import kotlinx.coroutines.test.runBlockingTest
+import mozilla.components.concept.storage.Login
+import mozilla.components.concept.storage.LoginValidationDelegate.Result
+import mozilla.components.concept.storage.LoginsStorage
+import mozilla.components.service.sync.logins.DefaultLoginValidationDelegate.Companion.mergeWithLogin
+import mozilla.components.support.test.any
+import mozilla.components.support.test.whenever
+import org.junit.Assert.assertEquals
+import org.junit.Before
+import org.junit.Test
+import org.junit.runner.RunWith
+import org.mockito.ArgumentMatchers.anyString
+import org.robolectric.annotation.Config
+
+@RunWith(AndroidJUnit4::class)
+@ExperimentalCoroutinesApi
+class DefaultLoginValidationDelegateTest {
+
+    private lateinit var loginsStorage: LoginsStorage
+    private lateinit var validationDelegate: DefaultLoginValidationDelegate
+    private lateinit var scope: TestCoroutineScope
+
+    @Before
+    @Config(sdk = [21])
+    fun before() {
+        loginsStorage = mockLoginsStorage()
```

Comment:
Let's use real storage implementation here (see `browser-storage-sync` for an example of how to hook that up.).
Your logic depends on behaviour of `ensureValid`, which you're not really testing below. Once you're using real storage implementation for these tests, the tests will actually be nicely end-to-end - you can insert real records to simulate these scenarios, and exercise our whole stack.

Once/if this code largely moves into Rust, test exercising our whole stack will remain valuable as integration tests. As they are now, they won't be testing much at that point.

What I'd like to see is a top-down approach here, working from the basic scenarios. Group the tests together so it's clear that we're testing all of these scenarios in an exhaustive way:
- various duplicate, invalid record scenarios
- various 'new record' scenarios. e.g. changed username, new record entirely, new username for an origin with an already saved login, etc.
- updated records - changed password, what else? which fields are considered immutable?